### DISCUSSION TITLE
RFC: Standardized + Gov-Friendly Collator Staking/Deposit through EVM+ Smart Contracts

### DISCUSSION THREAD

**Discussion by hackfisher** [darwinia-network/darwinia#1481] [General]
This proposal aims to reimplement Collator Staking and Deposit functionalities through EVM+ smart contracts to achieve two significant objectives:

1. Standardize Collator Staking and Deposit to ensure better compatibility with ERC20 and ERC721 standards. This integration will enhance our infrastructure and ecosystem interactions, for instance, making the system more accommodating to Liquidity Staking Tokens (LSTs). Additionally, it allows transforming deposits into NFTs, which can be transferred and traded in the [NFT market](https://github.com/darwinia-network/collaboration/pull/46).

2. As [prerequisite](https://github.com/orgs/darwinia-network/discussions/1284) to the next-generation governance system, transitioning Collator Staking and Deposit to smart contracts will facilitate the adoption of standardized governance models like those provided by [Tally](https://github.com/orgs/dcdao/discussions/127)/OpenZeppelin [Governor contracts](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/governance/Governor.sol). This shift will make on-chain governance more intelligent and standardized.

For Collator Staking, a critical step during runtime is the collator selection and ordering for new sessions, which can now be managed within a smart contract. At the beginning of a new session, the runtime pallet will directly retrieve information from the configured Collator Staking smart contract(Note: When implementing, fallback to previous/backup set is necessary if new set reading is empty or error).

Following [DIP-6](https://dips.darwinia.network/DIPs/dip-6.html), Collator Staking has been simplified; however, it still involves RING and Deposit components. To simplify implementation, encapsulating them within smart contracts is advantageous. Similar to [DIP-5](https://dips.darwinia.network/DIPs/dip-5.html), a migration phase of 30 to 60 days should be established to ensure a smooth transition, allowing users to migrate their existing staking to the new system.

More specifically, we will first implement a Collator Staking contract where users can stake/unstake RING and nominate collators.

Additionally, a Deposit contract will replicate the functionality of the original deposit module, where deposits are converted into an ERC721 NFT Token. Users can then use this NFT for further staking/unstaking in the Collator Staking contract.

Support for LSTs will align with standard practices in other ecosystems, such as implementing a Liquid RING Collator Staking contract. Users can engage in staking and unstaking through this contract, which will issue an ERC20 stRING token upon staking. This token can be transferred, traded, or used to add liquidity in DEXes. Developers within the ecosystem can also create customized versions.

Similarly, a Liquid Deposit Collator Staking version will be developed for deposit staking, allowing users to obtain an NFT voucher that can be transferred or traded while staking.

For participation in the next-generation governance system, the Liquid RING Collator Staking Token and Liquid Deposit Collator Staking Token will be highly conducive, merely requiring configuration of voting weights in the Governor contract.

Please have @AurevoirXavier @hujw77 @wuminzhe @boundless-forest attentions to this RFC and prepare for subsequent DIPs, engineering implementations, docs and education. Thank you.

**Comment by 2xJDubs** [comment]
I really like the direction of this RFC and I know many of us have been waiting to be able to vote with our staked and deposited RING.  I do have a couple questions and thoughts around what is being considered.

1. If deposits are going to be represented by ERC721 tokens, does that mean minting of KTON for deposits will stop?
2. If KTON minting stops will it still be used in staking or will there be a swap of KTON to these ERC721 tokens?
3. Would the liquid staking tokens, stRING, be used for governance or will there be a way to pole the staking and deposit contracts to get an accounts holding?

Very excited for this change over all.

**Reply by hackfisher** [reply]
> If deposits are going to be represented by ERC721 tokens, does that mean minting of KTON for deposits will stop?

No.

> If KTON minting stops will it still be used in staking or will there be a swap of KTON to these ERC721 tokens?

Not validate question, see above.

> Would the liquid staking tokens, stRING, be used for governance 

Not all the LST can be used for gov, only the LST tokens approved by gov can be used in governor contracts, which token can be used is also defined by governor contract and can be upgradable through gov.

> will there be a way to pole the staking and deposit contracts to get an accounts holding?

NO if I understand your question correctly, or unless there are security vulnerability exists.

**Comment by SasoLithops** [comment]
Love it. Governance participantion first. Second very good imho to prepare for liquid staked asset.

My questions:

1. Will it be possible to load NFT with tokens(Ring,Kton maybe others in future) so it could have an advantage in NFT marketplace?

2. _"More specifically, we will first implement a Collator Staking contract where users can stake/unstake RING and nominate collators."_

Will we now be able to nominate to more collators then one?

3. _"For participation in the next-generation governance system, the Liquid RING Collator Staking Token and Liquid Deposit Collator Staking Token will be highly conducive, merely requiring configuration of voting weights in the Governor contract."_

According to this paragraph there seem to be two tokens:

- The Liquid Ring Collator Staking Token ( is this stRING?)
- The Liquid Deposit Collator Staking Token ( is this the NFT?)

4. _"More specifically, we will first implement a Collator Staking contract where users can stake/unstake RING and nominate collators._

My question is how will this play out to already locked Ring? For example I have just locked a month ago some Ring for three years? I have over 1M Ring locked so will this automaticly be put under Collator staking contract?

5. _Additionally, a Deposit contract will replicate the functionality of the original deposit module, where deposits are converted into an ERC721 NFT Token."_

This one was already asked by Jesse. But we will still mint Kton for deposit lock? And time periods 1-36 months will remain?

Thank you for your hard work on this🏆

**Reply by hackfisher** [reply]
> 1. Will it be possible to load NFT with tokens(Ring,Kton maybe others in future) so it could have an advantage in NFT marketplace?

Yes, we are seeking DApps like NFT marketplace in collaboration repo, one is working in progress: https://github.com/darwinia-network/collaboration/blob/master/grant/projects/sublayer.md

> 2. Will we now be able to nominate to more collators then one?

Not decided but possible, to keep it simple, it is not required in design, but possible according to current progress of engineering detail draft.

> 3.
yes, the token contract might have a interface telling how many RING votes they have, this is still detail in drafting.

> 4.
> My question is how will this play out to already locked Ring? For example I have just locked a month ago some Ring for three years? I have over 1M Ring locked so will this automaticly be put under Collator staking contract?

Some user migration might be needed

> But we will still mint Kton for deposit lock? And time periods 1-36 months will remain?

This is not in the scope of this proposal, so it is unchanged unless other proposal introduce this.

**Comment by aurexav** [comment]
https://dips.darwinia.network/DIPs/dip-7.html
